snuba/query/processors/logical/filter_in_select_optimizer.py

```diff
@@ -1,234 +1,145 @@
 import logging
+from copy import deepcopy
 
 from snuba import environment
 from snuba.query.composite import CompositeQuery
 from snuba.query.conditions import binary_condition
 from snuba.query.data_source.simple import Entity as QueryEntity
 from snuba.query.expressions import (
+    Argument,
     Column,
     CurriedFunctionCall,
-    Expression,
+    ExpressionVisitor,
     FunctionCall,
+    Lambda,
     Literal,
     SubscriptableReference,
 )
 from snuba.query.logical import Query as LogicalQuery
 from snuba.state import get_int_config
 from snuba.utils.metrics.wrapper import MetricsWrapper
 
-"""
-Domain maps from a property to the specific values that are being filtered for. Ex:
-    sumIf(value, metric_id in [1,2,3])
-        Domain = {metric_id: {1,2,3}}
-    sumIf(value, metric_id=5 and status=200) / sumIf(value, metric_id=5 and status=400)
-        Domain = {
-            metric_id: {5},
-            status: {200, 400}
-        }
-"""
-Domain = dict[Column | SubscriptableReference, set[Literal]]
-
 metrics = MetricsWrapper(environment.metrics, "api")
 
 logger = logging.getLogger(__name__)
 
 
+class FindConditionalAggregateFunctionsVisitor(ExpressionVisitor[None]):
+    """
+    Search an expression for all conditional aggregate functions.
+
+    Might be able to use matchers.Pattern to extend this idea for general search but I was too lazy.
+    """
+
+    def __init__(self) -> None:
+        self._matches: list[FunctionCall | CurriedFunctionCall] = []
+
+    def get_matches(self) -> list[FunctionCall | CurriedFunctionCall]:
+        return deepcopy(self._matches)
+
+    def visit_literal(self, exp: Literal) -> None:
+        return
+
+    def visit_column(self, exp: Column) -> None:
+        return
+
+    def visit_subscriptable_reference(self, exp: SubscriptableReference) -> None:
+        return
+
+    def visit_function_call(self, exp: FunctionCall) -> None:
+        if exp.function_name[-2:] == "If":
+            self._matches.append(exp)
+        else:
+            for param in exp.parameters:
+                param.accept(self)
+
+    def visit_curried_function_call(self, exp: CurriedFunctionCall) -> None:
+        if exp.internal_function.function_name[-2:] == "If":
+            self._matches.append(exp)
+
+    def visit_argument(self, exp: Argument) -> None:
+        return
+
+    def visit_lambda(self, exp: Lambda) -> None:
+        return
+
+
 class FilterInSelectOptimizer:
```

Comment:
This class is a query processor it should extend the query processor class

Reply:
The main issue I ran into trying to do this is:
LogicalQueryProcessor defines input as a logical query, but the query I will be given by the parser is logical or composite

Smaller (i think) issue I ran into is:
My processor is being called with the _post_process function which wont pass in query settings as required by LogicalQueryProcessor

Reply:
We mentioned this optimizer being in snuba/query/processors/logical is a bit weird. Maybe I could move it to snuba/query/mql, any other ideas?

Reply:
how does this processor handle composite queries? Did you test for that?

Reply:
For this processor we can assume that it will never be used for Composite queries. It's meant to solve a very specific problem for MQL, and CompositeQuery support is out of scope.

So what you could do is simply do nothing if the query is a CompositeQuery. Then the typing will match.

Reply:
made it logical processor